### PR TITLE
Scheduler refactor

### PR DESCRIPTION
**Pull request by alexgparity**

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2118665

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2118923

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2119904

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2122169

File: runtime/parachains/src/scheduler.rs [hunk 1]
```diff
@@ -386,37 +378,46 @@ impl<T: Config> Pallet<T> {
...
+	) -> Option<ParathreadEntry> {
+		let idx = freed_idx.0 as usize;
+		if idx < cores.len() {
+			match cores[idx].take() {
+				None | Some(CoreOccupied::Parachain) => None,
+				Some(CoreOccupied::Parathread(entry)) => Some(entry),
+			}
+		} else {
+			None
+		}
```
[the remark this answers is not in the document]
Comment (alexgparity) [reply runtime/parachains/src/scheduler.rs:418]: Nope, doesn't work. I tried a few different things here but they all sucked.

[the remark this answers is not in the document]
Comment (sandreim) [reply runtime/parachains/src/scheduler.rs:418]: Ah, sorry. Corrected below and should compile fine:

```
		match cores.get_mut(idx).take().cloned() {
			None | Some(None) | Some(Some(CoreOccupied::Parachain)) => None,
			Some(Some(CoreOccupied::Parathread(entry))) => Some(entry),
		}
```

[the remark this answers is not in the document]
Comment (alexgparity) [reply runtime/parachains/src/scheduler.rs:418]: I tried that. With an added `.flatten()` to remove one layer of `Option`. That breaks tests though. I don't really understand why but I don't want to spend time on that for now. If you want to have a look, `scheduler::tests::schedule_schedules_including_just_freed` breaks for me locally with the above.

**Pull request closed without merging**